pkg/webhook/sidecar_spec.go

```diff
@@ -118,6 +113,95 @@ func GetSidecarContainerSpec(c *Config) corev1.Container {
 	return container
 }
 
+// GetSecurityContext ensures the sidecar that uses it follows Restricted Pod Security Standard.
+// See https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted
+func GetSecurityContext() *corev1.SecurityContext {
+	return &corev1.SecurityContext{
+		AllowPrivilegeEscalation: ptr.To(false),
+		ReadOnlyRootFilesystem:   ptr.To(true),
+		Capabilities: &corev1.Capabilities{
+			Drop: []corev1.Capability{
+				corev1.Capability("ALL"),
+			},
+		},
+		SeccompProfile: &corev1.SeccompProfile{Type: corev1.SeccompProfileTypeRuntimeDefault},
+		RunAsNonRoot:   ptr.To(true),
+		RunAsUser:      ptr.To(int64(NobodyUID)),
+		RunAsGroup:     ptr.To(int64(NobodyGID)),
+	}
+}
+
+func (si *SidecarInjector) GetNativeMetadataPrefetchSidecarContainerSpec(pod *corev1.Pod, image string) corev1.Container {
+	container := si.GetMetadataPrefetchSidecarContainerSpec(pod, image)
+	container.Env = append(container.Env, corev1.EnvVar{Name: "NATIVE_SIDECAR", Value: "TRUE"})
+	container.RestartPolicy = ptr.To(corev1.ContainerRestartPolicyAlways)
+
+	return container
+}
+
+func getMetadataPrefetchConfig(image string) *Config {
+	return &Config{
+		CPURequest:              resource.MustParse("10m"),
+		CPULimit:                resource.MustParse("50m"),
+		MemoryRequest:           resource.MustParse("10Mi"),
+		MemoryLimit:             resource.MustParse("10Mi"),
+		EphemeralStorageRequest: resource.MustParse("10Mi"),
+		EphemeralStorageLimit:   resource.MustParse("10Mi"),
+		MetadataContainerImage:  image,
+	}
+}
+
+func (si *SidecarInjector) GetMetadataPrefetchSidecarContainerSpec(pod *corev1.Pod, image string) corev1.Container {
+	if pod == nil {
+		klog.Warning("failed to get metadata prefetch container spec: pod is nil")
+
+		return corev1.Container{}
+	}
+
+	c := getMetadataPrefetchConfig(image)
+	limits, requests := prepareResourceList(c)
+
+	// The sidecar container follows Restricted Pod Security Standard,
+	// see https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted
+	container := corev1.Container{
+		Name:            MetadataPrefetchSidecarName,
+		Image:           c.MetadataContainerImage,
+		ImagePullPolicy: corev1.PullPolicy(c.ImagePullPolicy),
+		SecurityContext: GetSecurityContext(),
+		Resources: corev1.ResourceRequirements{
+			Limits:   limits,
+			Requests: requests,
+		},
+		VolumeMounts: []corev1.VolumeMount{},
+	}
+
+	for _, v := range pod.Spec.Volumes {
+		isGcsFuseCSIVolume, volumeAttributes, err := si.isGcsFuseCSIVolume(v, pod.Namespace)
+		if err != nil {
+			klog.Errorf("failed to determine if %s is a GcsFuseCSI backed volume: %v", v.Name, err)
```

Comment:
seem to me like we will write error logs for volumes which are not of type gcsfusecsi (https://github.com/GoogleCloudPlatform/gcs-fuse-csi-driver/pull/360/files#diff-b6aadbfe5e9536eb570a91f11fea0ba3127177b43e0b865f2389ba9540bebe33R68). This error log can be misleading during debugging.


Any volume for which csidriver name is not the same can simply be skipped in the  GetPreprovisionCSIVolume (similar to how you are skipping it for csi ephemeral volume which does not match the gcsfuse csidriver name https://github.com/GoogleCloudPlatform/gcs-fuse-csi-driver/pull/360/files#diff-838b88d01a3358247a8414ffe005610bd045beca211c41cb74ad9996a984a832R36)

Reply:
Seems that this is the only place we are logging failures. This log is only run when we return an error from `GetPreprovisionCSIVolume`. I modified `GetPreprovisionCSIVolume` to only return errors when it is obvious that there's a misconfiguration in the Volume, instead of when the volume type is gcsfusecsi.